chips/nrf51/src/radio.rs

```diff
@@ -0,0 +1,456 @@
+    #[no_mangle]
+    pub fn handle_interrupt(&self) {
+        let regs = unsafe { &*self.regs };
+        self.disable_nvic();
+        self.disable_interrupts();
+        nvic::clear_pending(NvicIdx::RADIO);
+
+        if regs.READY.get() == 1 {
+            if regs.STATE.get() <= 4 {
+                self.set_rx_buffer();
+            } else {
+                self.set_tx_buffer();
+            }
+            regs.READY.set(0);
+            regs.END.set(0);
+            regs.START.set(1);
+        }
+
+        if regs.PAYLOAD.get() == 1 {
+            regs.PAYLOAD.set(0);
+        }
+
+        if regs.ADDRESS.get() == 1 {
+            regs.ADDRESS.set(0);
+        }
+
+        if regs.END.get() == 1 {
+            regs.END.set(0);
+            regs.DISABLE.set(1);
+            // this state only verifies that END is received in TX-mode
+            // which means that the transmission is finished
+            match regs.STATE.get() {
+                RADIO_STATE_TXRU |
+                RADIO_STATE_TXIDLE |
+                RADIO_STATE_TXDISABLE |
+                RADIO_STATE_TX => {
+                    match regs.FREQEUNCY.get() {
+                        // frequency 39
+                        80 => {
+                            self.radio_off();
+                        }
+                        // frequency 38
+                        26 => {
+                            self.set_channel_freq(39);
+                            self.set_data_white_iv(39);
+                            regs.READY.set(0);
+                            regs.TXEN.set(1);
+                        }
+                        // frequency 37
+                        2 => {
+                            self.set_channel_freq(38);
+                            self.set_data_white_iv(38);
+                            regs.READY.set(0);
+                            regs.TXEN.set(1);
+                        }
+                        // don't care as we only support advertisements at the moment
+                        _ => {
+                            self.set_channel_freq(37);
+                            self.set_data_white_iv(37)
+                        }
+                    }
+                }
+                _ => (),
+            }
+        }
+        self.enable_nvic();
+        self.enable_interrupts();
+    }
+
+
+    pub fn enable_interrupts(&self) {
+        // INTENSET REG
+        let regs = unsafe { &*self.regs };
+        // 15 i.e set 4 LSB
+        regs.INTENSET.set(1 | 1 << 1 | 1 << 2 | 1 << 3);
+    }
+
+    pub fn disable_interrupts(&self) {
+        let regs = unsafe { &*self.regs };
+        // disable all possible interrupts
+        regs.INTENCLR.set(0x4ff);
+    }
+
+    pub fn enable_nvic(&self) {
+        nvic::enable(NvicIdx::RADIO);
+    }
+
+    pub fn disable_nvic(&self) {
+        nvic::disable(NvicIdx::RADIO);
+    }
+
+    pub fn reset_payload(&self) {
+        // reset contents except header || address
+        for i in 9..39 {
```

Comment:
I think it's still preferable here. (`for i in 9..(unsafe { PAYLOAD.len()})`). There is a cleaner (and overall safer) way to store `PAYLOAD` which would not require all these unsafes in general, but I think that can wait for another pass after this is merged.